proof-of-spacetime.md

```diff
@@ -0,0 +1,187 @@
+## Proof-of-Spacetime
+
+This document describes Rational-PoSt, the Proof-of-Spacetime used in Filecoin.
+
+## Rational PoSt
+
+### Definitions
+
+- **POST_PROVING_PERIOD**: The time interval in which a PoSt has to be submitted.
+- **POST_CHALLENGE_TIME**: The time offset at which the actual work of generating the PoSt should be started. This is some delta before the end of the `Proving Period`, and as such less then a single `Proving Period`.
+
+### Execution Flow
+
+```
+    ■──────────────────────── Proving Period ─────────────────────────■
+
+
+    ▲───────────────────────────────────────────────────▲─────────────▲
+    │                                                   │             │
+    │                                                   │             │
+    │                                                   │             │
+┌───────┐                                     ┌──────────────────┐┌───────┐
+│ Start │                                     │ Generation Start ││  End  │
+└───────┘                                     └──────────────────┘└───────┘
+                                                        ▲
+
+                                                        │
+                                                                       ┌ ─ ─ ─ ─ ─ ─ ─ ─ ┐
+                                                        └ ─ ─ ─ ─ ─ ─ ─ Randomness Input
+                                                                       └ ─ ─ ─ ─ ─ ─ ─ ─ ┘
+```
+
+TODO: Add post submission to the diagram.
+
+
+### High Level API
+
+#### Fault Detection
+
+Fault detection happens over the course of the life time of a sector. When the sector is for some reason unavailable, the miner is responsible to post an up to date `AddFaults` message to the chain. When recovering any faults, they need to submit a `RecoverFaults` message. The PoSt generation then takes the latest available `faults` of the miner to generate a PoSt matching the committed sectors and faults.
+
+TODO: Is it okay to add a fault and recover from it in the same proving period?
+
+#### Generation
+
+```go
+func GeneratePoSt(sectorSize BytesAmount, sectors []commR, seed []byte, faults FaultSet) PoStProof {
+    // Generate the Merkle Inclusion Proofs + Faults
+
+    inclusionProofs := []
+    challenges := DerivePoStChallenges(sectorSize, seed, faults)
+
+    for n in 0..POST_CHALLENGES_COUNT {
+        challenge := challenges[n]
+        sector := challenge % len(sectors)
+
+        // Leaf index of the selected sector
+        challenge_value = challenge % sectorSize
```

Comment:
Is it really safe to derive the challenged sector and challenged leaf in this way? This seems to heavily correlate the two. For example, suppose `len(sectors)` and `sector_size / 32` were the same. (I know this is unrealistic.) In that case, the challenged leaf would be entirely determined by the challenged sector (for a given set of sectors). That means a miner whose sectors are static (at capacity) would always receive the same challenges for each sector and could simply cache the proof for one leaf (no data storage required). 

This is obviously an extreme case, but its existence is troubling. A more plausible example: suppose a miner is storing 1TiB, with a sector size of 1GiB. That would make `len(sectors) = 1024` and `sector_size = 1073741824 `, with one being 1048576 times the other. In this case, it seems that there are only 1048576 possible challenged leafs for any given sector. If true, that would allow storage of only these 1048576 proofs (each 32x25 = 800 bytes) for a 22% space reduction.

As far as I can tell, this issue exists any time the sector size is a multiple of the of the number of sectors being proved. If this is true, rather than come up with clever ways to reduce the risk, I think we should probably use entirely independent challenge values to select the sector and the leaf within that sector. (There are other strategies, but I think what I propose is simple and safe.)

To see why the current challenge derivation method is problematic, consider:

When SECTOR_SIZE = M * SECTOR_COUNT, then for any given challenge C such that C % SECTOR_COUNT = CHOSEN_SECTOR, there are only M distinct values of C % SECTOR_SIZE. The possible chosen leaves are: i * M + CHOSEN_SECTOR for i in [1..M]. Any C which yields a different leaf when selecting by C % SECTOR_SIZE will also yield a different sector when selecting by C % SECTOR_COUNT.

Make sense?

cc: @nicola @nikkolasg

Reply:
proposal: given that we already have 32bytes of randomness and only need 64bits of it at the moment we can do 
- take the low 8 bytes to select the sector
- take the high 8 bytes to select the leaf

thoughts?

Reply:
It might be a little cleaner just to take the first 8 bytes for one and the next 8 bytes for the other, but either way would address this issue.

Reply:
I have switched to this, makes it clearer how challenge derivation works as well. I am not a 100% sure it is an issue, but the new solution should be fully safe

Reply:
Thanks, @dignifiedquire.